File: Array/4Sum II/solution.go
package main

/*
A+B+C+D = 0 <=> A+B = -C-D
*/

func fourSumCount(nums1 []int, nums2 []int, nums3 []int, nums4 []int) int {
	n := len(nums1)
	result := 0
	sumMap := make(map[int]int)
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			sumMap[nums1[i]+nums2[j]]++
		}
	}
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			sum := nums3[i] + nums4[j]
			if mutiplier, ok := sumMap[-sum]; ok {
				result += mutiplier
			}
		}
	}
	return result
}

File: Array/4Sum II/solution.py
class Solution:
    def fourSumCount(self, nums1: List[int], nums2: List[int], nums3: List[int], nums4: List[int]) -> int:
        sums = Counter()
        result = 0
        for num1 in nums1:
            for num2 in nums2:
                sums[num1 + num2] += 1
        
        for num3 in nums3:
            for num4 in nums4:
                result += sums[-(num3 + num4)]
        
        return result
